datarobot_provider/hooks/connections.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2022 DataRobot, Inc. and its affiliates.
+#
+# All rights reserved.
+#
+# This is proprietary source code of DataRobot, Inc. and its affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+from typing import Any
+from typing import Dict
+
+from airflow import AirflowException
+from airflow.hooks.base import BaseHook
+from datarobot.models.data_store import DataStore
+from datarobot.models.driver import DataDriver
+
+from datarobot_provider.hooks.datarobot import DataRobotHook
+
+
+class JDBCDataSourceHook(BaseHook):
+    """
+    A hook that interacts with DataRobot via its public Python API library to
+    manage JDBC connections with corresponding credentials.
+
+    :param datarobot_jdbc_conn_id: Connection ID, defaults to `datarobot_jdbc_default`
+    :type datarobot_jdbc_conn_id: str, optional
+    """
+
+    conn_name_attr = 'datarobot_jdbc_conn_id'
+    default_datarobot_jdbc_conn_name = 'datarobot_jdbc_default'
+    conn_type = 'datarobot_jdbc_datasource'
+    hook_name = 'DataRobot JDBC DataSource'
+
+    @staticmethod
+    def get_connection_form_widgets() -> Dict[str, Any]:
+        """Returns connection widgets to add to connection form."""
+        from flask_appbuilder.fieldwidgets import BS3TextFieldWidget
+        from flask_babel import lazy_gettext
+        from wtforms import StringField
+
+        return {
+            "datarobot_connection": StringField(
+                lazy_gettext('DataRobot Connection'),
+                widget=BS3TextFieldWidget(),
+                default='datarobot_default',
+            ),
+            "jdbc_driver": StringField(
+                lazy_gettext('JDBC Driver'),
+                widget=BS3TextFieldWidget(),
+                default='',
+            ),
+            "jdbc_url": StringField(
+                lazy_gettext('JDBC URL'),
+                widget=BS3TextFieldWidget(),
+                default='jdbc:',
+            ),
+        }
+
+    @staticmethod
+    def get_ui_field_behaviour() -> Dict:
+        """Returns custom field behaviour."""
+        return {
+            "hidden_fields": ['host', 'schema', 'port', 'extra'],
+            "relabeling": {},
+            "placeholders": {
+                'datarobot_connection': 'datarobot_default',
+                'jdbc_driver': '',
+                'jdbc_url': 'jdbc:',
+                'login': '',
+                'password': '',
```

Comment:
Do all jdbc connections have a username and password? I thought you can encode username and password values directly in the connection URL?

Reply:
thats because of DataRobot API, its expecting username and password as a separate credentials_data object. Also its not secure to keep password as a part of connection string and sometimes it might cause problems with special characters inside a password. So I would suggest to force users to use login&password as a separate fields